python/sparkdl/image/imageIO.py

```diff
@@ -26,90 +25,55 @@
 from pyspark import SparkContext
 from pyspark.sql.types import (BinaryType, IntegerType, StringType, StructField, StructType)
 from pyspark.sql.functions import udf
+from pyspark.ml.image import ImageSchema
 
 
-imageSchema = StructType([StructField("mode", StringType(), False),
-                          StructField("height", IntegerType(), False),
-                          StructField("width", IntegerType(), False),
-                          StructField("nChannels", IntegerType(), False),
-                          StructField("data", BinaryType(), False)])
-
-
-# ImageType class for holding metadata about images stored in DataFrames.
+# ImageType represents supported OpenCV types
 # fields:
+#   name - OpenCvMode
+#   ord - Ordinal of the corresponding OpenCV mode (stored in mode field of ImageSchema).
 #   nChannels - number of channels in the image
-#   dtype - data type of the image's "data" Column, sorted as a numpy compatible string.
-#   channelContent - info about the contents of each channel currently only "I" (intensity) and
-#     "RGB" are supported for 1 and 3 channel data respectively.
-#   pilMode - The mode that should be used to convert to a PIL image.
-#   sparkMode - Unique identifier string used in spark image representation.
-ImageType = namedtuple("ImageType", ["nChannels",
-                                     "dtype",
-                                     "channelContent",
-                                     "pilMode",
-                                     "sparkMode",
-                                     ])
-class SparkMode(object):
-    RGB = "RGB"
-    FLOAT32 = "float32"
-    RGB_FLOAT32 = "RGB-float32"
-
-supportedImageTypes = [
-    ImageType(3, "uint8", "RGB", "RGB", SparkMode.RGB),
-    ImageType(1, "float32", "I", "F", SparkMode.FLOAT32),
-    ImageType(3, "float32", "RGB", None, SparkMode.RGB_FLOAT32),
-]
-pilModeLookup = {t.pilMode: t for t in supportedImageTypes
-                 if t.pilMode is not None}
-sparkModeLookup = {t.sparkMode: t for t in supportedImageTypes}
-
-
-def imageArrayToStruct(imgArray, sparkMode=None):
+#   dtype - data type of the image's array, sorted as a numpy compatible string.
+OcvType = namedtuple("OcvType",["name","ord","nChannels","dtype"])
+
+supportedOcvTypes = (
+    OcvType(name="CV_8UC1",  ord=0,  nChannels=1, dtype="uint8"  ),
+    OcvType(name="CV_32FC1", ord=5,  nChannels=1, dtype="float32"),
+    OcvType(name="CV_8UC3",  ord=16, nChannels=3, dtype="uint8"  ),
+    OcvType(name="CV_32FC3", ord=21, nChannels=3, dtype="float32"),
+    OcvType(name="CV_8UC4",  ord=24, nChannels=4, dtype="uint8"  ),
+    OcvType(name="CV_32FC4", ord=29, nChannels=4, dtype="float32"),
+)
+
+__ocvTypesByName = {m.name:m for m in supportedOcvTypes}
+__ocvTypesByOrdinal = {m.ord:m for m in supportedOcvTypes}
+
+def imageTypeByOrdinal(ord):
```

Comment:
Not sure if it should be private, we call from other files. I think it is actually a useful function for the ImageSchema.

Reply:
Those basic classes seem right. I think the two main ways that user-facing APIs are exposed in python are by either:

1) Promoting them to a top level (like numpy) or API (like matplotlib.pyplot) namespace.
2) Through documentation. I think some tools use `__all__` to identify which functions and classes to auto-generate docs for so that could work nicely but I have to check.

I agree that we should avoid importing `_private` members from a module. My default is usually to identify things as "internal" unless I explicitly want to warn other developers not use them or they're meant to be public APIs. That's a bit of a personal preference and we could set a style guideline here to some level of consistency around this.

On somewhat unrelated note we should avoid using names with two underscores, eg `__ptyhon_private`, unless we're explicitly trying to invoke python's name mangling, https://en.wikipedia.org/wiki/Name_mangling#Python.